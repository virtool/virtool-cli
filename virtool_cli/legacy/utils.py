from dataclasses import dataclass
from pathlib import Path
from typing import Generator

import orjson
from orjson import orjson
from pydantic_core import ErrorDetails

from virtool_cli.ncbi.client import NCBIClient


class HandleErrorContext:
    """A context that is passed to a ``ValidationError`` error handler.

    This context is for one error from ``ValidationError.errors``, not for the
    ``ValidationError`` itself.

    """

    def __init__(
        self,
        error: ErrorDetails,
        fix: bool,
        ncbi_client: NCBIClient,
        otu: dict,
        repaired_otu: dict,
    ):
        self.error = error
        """The details of the Pydantic validation error."""

        self.fix = fix
        """Whether to attempt to fix the error or not."""

        self.ncbi_client = ncbi_client
        """An instance of the NCBI client."""

        self.otu = otu
        """A dictionary representation of the OTU."""

        self.repaired_otu = repaired_otu
        """A copy of the OTU that repairs should be applied to."""

    def update_otu(self, update: dict):
        """Update OTU associated with the error.

        Example:
        -------
        .. code-block:: python

            ctx.update_otu({"name": "new_name"})

        :param update: the update dict

        """
        self.repaired_otu.update(update)

    def update_isolate(self, isolate_index: int, update: dict):
        """Update a legacy isolate on disk given an update dict.

        Example:
        -------
        .. code-block:: python

           ctx.update_isolate(
               isolate_index,
              {"source_name": source.name, "source_type": source.type.lower()},
           )

        :param isolate_index: the index of the isolate in the OTU
        :param update: the update dict

        """
        self.repaired_otu["isolates"][isolate_index].update(update)

    def update_sequence(self, isolate_index: int, sequence_index: int, update: dict):
        """Update a legacy sequence on disk given an update dict.

        Example:
        -------
        .. code-block:: python

            ctx.update_sequence(
                isolate_index,
                sequence_index,
                {"sequence": "ATGCATCGTATGCTYACTAGCTAC"},
            )

        """
        self.repaired_otu["isolates"][isolate_index]["sequences"][
            sequence_index
        ].update(
            update,
        )


@dataclass
class ErrorHandledResult:
    """The return value for a legacy validation ``ValidationError`` handler."""

    message: str
    """The message to display to the user.

    Includes ``rich``-compatible markup.
    """

    fixed: bool = False
    """Whether the error was fixed.

    This value is used to show the user whether the error was fixed or not.
    """


def build_legacy_otu(path: Path) -> dict:
    """Build a dict-based representation of an OTU from an OTU directory.

    :param path: the path to the OTU directory
    :return: a dict-based representation of the OTU

    """
    with open(path / "otu.json") as f:
        otu_data = orjson.loads(f.read())

    otu_data["isolates"] = []

    for isolate_dir_path in sorted(path.iterdir()):
        if isolate_dir_path.name == "otu.json":
            continue

        with open(isolate_dir_path / "isolate.json") as f:
            isolate_data = orjson.loads(f.read())

        isolate_data["sequences"] = []

        for sequence_path in sorted(isolate_dir_path.iterdir()):
            if sequence_path.name == "isolate.json":
                continue

            with open(sequence_path) as f:
                sequence_data = orjson.loads(f.read())

            isolate_data["sequences"].append(sequence_data)

        otu_data["isolates"].append(isolate_data)

    return otu_data


def replace_otu(path: Path, otu: dict):
    """Replace the otu in the path with the data in ``otu``.

    :param path: the path to the otu
    :param otu: the new otu data

    """
    if not path.parent.is_dir():
        raise FileNotFoundError(f"No such OTU parent directory: {path.parent}")

    if not path.is_dir():
        raise FileNotFoundError(f"No such OTU directory: {path}")

    with open(path / "otu.json", "wb") as f:
        f.write(
            orjson.dumps(
                {
                    "_id": otu["_id"],
                    "name": otu["name"],
                    "abbreviation": otu["abbreviation"],
                    "schema": otu["schema"],
                    "taxid": otu["taxid"],
                },
                option=orjson.OPT_INDENT_2,
            ),
        )

    for isolate in sorted(otu["isolates"], key=lambda x: x["id"]):
        isolate_id = isolate["id"]

        with open(path / isolate_id / "isolate.json", "wb") as f:
            f.write(
                orjson.dumps(
                    {
                        "id": isolate["id"],
                        "default": isolate["default"],
                        "source_name": isolate["source_name"],
                        "source_type": isolate["source_type"],
                    },
                    option=orjson.OPT_INDENT_2,
                ),
            )

        for sequence in sorted(isolate["sequences"], key=lambda x: x["_id"]):
            with open(path / isolate_id / f"{sequence['_id']}.json", "wb") as f:
                f.write(
                    orjson.dumps(
                        {
                            "_id": sequence["_id"],
                            "accession": sequence["accession"],
                            "definition": sequence["definition"],
                            "host": sequence["host"],
                            "segment": sequence["segment"],
                            "sequence": sequence["sequence"],
                        },
                        option=orjson.OPT_INDENT_2,
                    ),
                )


def iter_legacy_otus(src_path: Path) -> Generator[dict, None, None]:
    """Iterate over all OTUs in a legacy repository.

    :param src_path: The path to the legacy repository.
    :return: a generator that yields OTU data dictionaries.
    """
    for alpha_dir_path in sorted(src_path.iterdir()):
        if alpha_dir_path.name == "meta.json":
            continue

        for otu_dir_path in sorted(alpha_dir_path.iterdir()):
            yield build_legacy_otu(otu_dir_path)
