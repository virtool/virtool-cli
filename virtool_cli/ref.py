from pathlib import Path

import click

import virtool_cli.build
import virtool_cli.divide
import virtool_cli.isolate
import virtool_cli.repair
import virtool_cli.taxid


@click.group("ref")
def ref():
    """Commands related to references"""
    pass


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a database reference directory",
)
@click.option(
    "-o",
    "--output",
    default="reference.json",
    help="the output path for a reference.json file",
)
@click.option("-i", "--indent", is_flag=True)
@click.option(
    "-V",
    "--version",
    default=None,
    type=str,
    help="the version string to include in the reference.json file",
)
def build(src_path, output, indent, version):
    """Build a Virtool reference JSON file from a reference directory."""
    try:
        virtool_cli.build.run(Path(src_path), Path(output), indent, version)
    except (FileNotFoundError, NotADirectoryError):
        click.echo("Not a valid reference directory")


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a input reference.json file",
)
@click.option(
    "-o",
    "--output",
    default="src",
    type=str,
    help="the output path for a divided reference directory tree",
)
def divide(src_path, output):
    """Divide a reference.json file from Virtool into a reference directory tree."""
    try:
        if not src_path.endswith(".json"):
            raise TypeError
        virtool_cli.divide.run(Path(src_path), Path(output))
    except (TypeError, FileNotFoundError):
        click.echo(
            "Specified reference file either does not exist or is not a proper JSON file"
        )


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to the input reference.json file",
)
@click.option("-f", "--force_update", is_flag=True)
def taxid(src_path, force_update):
    """Fetch taxid for all OTU in given reference directory."""
    try:
        virtool_cli.taxid.run(Path(src_path), force_update)
    except (FileNotFoundError, NotADirectoryError):
        click.echo("Not a valid reference directory")


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
def isolate(src_path):
    """Fetch new isolates for all OTU in a given reference directory."""
    try:
        virtool_cli.isolate.run(Path(src_path))
    except (FileNotFoundError, NotADirectoryError):
        click.echo("Not a valid reference directory")


@ref.command()
@click.option(
    "-src",
    "--src_path",
    required=True,
    type=str,
    help="the path to a reference directory",
)
def repair(src_path):
    """Fix every OTU in a given reference directory."""
    try:
        virtool_cli.repair.run(Path(src_path))
    except (FileNotFoundError, NotADirectoryError):
        click.echo("Not a valid reference directory")


if __name__ == "__main__":
    ref()
