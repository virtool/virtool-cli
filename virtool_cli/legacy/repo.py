"""Repo level checks for legacy repositories."""

from pathlib import Path

from virtool_cli.legacy.utils import iter_legacy_otus
from virtool_cli.utils.console import console


def check_unique_accessions(path: Path):
    """Validate that all accession fields are unique."""
    accessions = set()

    for otu in iter_legacy_otus(path / "src"):
        for isolate in otu["isolates"]:
            for sequence in isolate["sequences"]:
                if accession := sequence.get("accession"):
                    if accession in accessions:
                        console.log(f"Found non-unique accession: {accession}")
                    else:
                        accessions.add(accession)


def check_unique_ids(path: Path):
    """Validate that resource IDs are unique.

    * All OTU _id fields are unique.
    * All isolate id fields are unique.
    * All sequence _id fields are unique.
    """
    otu_ids = []
    isolate_ids = []
    sequence_ids = []

    for otu in iter_legacy_otus(path / "src"):
        otu_ids.append(otu["_id"])

        for isolate in otu["isolates"]:
            isolate_ids.append(isolate["id"])

            for sequence in isolate["sequences"]:
                sequence_ids.append(sequence["_id"])

    if len(otu_ids) != len(set(otu_ids)):
        console.log("Found non-unique OTU ids")

    if len(isolate_ids) != len(set(isolate_ids)):
        console.log("Found non-unique isolate ids")

    if len(sequence_ids) != len(set(sequence_ids)):
        console.log("Found non-unique sequence ids")


def check_unique_otu_abbreviations_and_names(path: Path):
    """Validate that OTU abbreviations and names are unique."""
    abbreviations = set()
    names = set()

    for otu in iter_legacy_otus(path / "src"):
        abbreviation = otu["abbreviation"]

        if abbreviation:
            if abbreviation in abbreviations:
                console.log(f"Found non-unique OTU abbreviation: {abbreviation}")
            else:
                abbreviations.add(abbreviation)

        name = otu["name"].lower()

        if name in names:
            console.log(f"Found non-unique OTU name: {name}")
        else:
            names.add(name)
