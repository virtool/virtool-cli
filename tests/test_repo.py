from pathlib import Path
from uuid import UUID

import orjson
import pytest

from virtool_cli.ref.repo import EventSourcedRepo
from virtool_cli.ref.resources import (
    EventSourcedRepoIsolate,
    EventSourcedRepoOTU,
    EventSourcedRepoSequence,
)
from virtool_cli.ref.utils import (
    DataType,
    IsolateName,
    Molecule,
    MolType,
    Strandedness,
    Topology,
)


@pytest.fixture
def initialized_repo(empty_repo: EventSourcedRepo):
    otu = empty_repo.create_otu(
        "TMV",
        None,
        "Tobacco mosaic virus",
        Molecule(Strandedness.SINGLE, MolType.RNA, Topology.LINEAR),
        [],
        12242,
    )

    isolate_a = empty_repo.create_isolate(otu.id, None, "A", "isolate")
    empty_repo.create_sequence(
        otu.id,
        isolate_a.id,
        "TMVABC",
        "TMV",
        None,
        "RNA",
        "ACGT",
    )

    yield empty_repo


def init_otu(empty_repo: EventSourcedRepo) -> EventSourcedRepoOTU:
    return empty_repo.create_otu(
        "TMV",
        "abcd1234",
        "Tobacco mosaic virus",
        Molecule(Strandedness.SINGLE, MolType.RNA, Topology.LINEAR),
        [],
        12242,
    )


def test_new(empty_repo: EventSourcedRepo, tmp_path: Path):
    """Test that creating a new ``Repo`` object returns the expected object and creates
    the expected directory structure.
    """
    assert empty_repo.path == tmp_path / "test_repo"
    assert empty_repo.last_id == 1

    assert empty_repo.meta.data_type == DataType.GENOME
    assert empty_repo.meta.name == "Generic Viruses"
    assert empty_repo.meta.organism == "virus"


class TestCreateOTU:
    def test_ok(self, empty_repo: EventSourcedRepo):
        """Test that creating an OTU returns the expected ``RepoOTU`` object and creates
        the expected event file.
        """
        otu = empty_repo.create_otu(
            "TMV",
            "abcd1234",
            "Tobacco mosaic virus",
            Molecule(Strandedness.SINGLE, MolType.RNA, Topology.LINEAR),
            [],
            12242,
        )

        assert otu == EventSourcedRepoOTU(
            id=otu.id,
            acronym="TMV",
            excluded_accessions=set(),
            isolates=[],
            legacy_id="abcd1234",
            name="Tobacco mosaic virus",
            molecule=Molecule(Strandedness.SINGLE, MolType.RNA, Topology.LINEAR),
            schema=[],
            taxid=12242,
        )

        with open(empty_repo.path.joinpath("src", "00000002.json")) as f:
            event = orjson.loads(f.read())

        del event["timestamp"]

        assert event == {
            "data": {
                "id": str(otu.id),
                "acronym": "TMV",
                "legacy_id": "abcd1234",
                "name": "Tobacco mosaic virus",
                "molecule": {
                    "strandedness": "single",
                    "type": "RNA",
                    "topology": "linear",
                },
                "rep_isolate": None,
                "schema": [],
                "taxid": 12242,
            },
            "id": 2,
            "query": {
                "otu_id": str(otu.id),
            },
            "type": "CreateOTU",
        }

        assert empty_repo.last_id == 2

    def test_duplicate_name(self, empty_repo: EventSourcedRepo):
        """Test that creating an OTU with a name that already exists raises a
        ``ValueError``.
        """
        empty_repo.create_otu(
            "TMV",
            None,
            "Tobacco mosaic virus",
            Molecule(Strandedness.SINGLE, MolType.RNA, Topology.LINEAR),
            [],
            12242,
        )

        with pytest.raises(
            ValueError,
            match="An OTU with the name 'Tobacco mosaic virus' already exists",
        ):
            empty_repo.create_otu(
                "TMV",
                None,
                "Tobacco mosaic virus",
                Molecule(Strandedness.SINGLE, MolType.RNA, Topology.LINEAR),
                [],
                12242,
            )

    def test_duplicate_legacy_id(self, empty_repo: EventSourcedRepo):
        """Test that creating an OTU with a legacy ID that already exists raises a
        ``ValueError``.
        """
        empty_repo.create_otu(
            "TMV",
            "abcd1234",
            "Tobacco mosaic virus",
            Molecule(Strandedness.SINGLE, MolType.RNA, Topology.LINEAR),
            [],
            12242,
        )

        with pytest.raises(
            ValueError,
            match="An OTU with the legacy ID 'abcd1234' already exists",
        ):
            empty_repo.create_otu(
                "",
                "abcd1234",
                "Abaca bunchy top virus",
                Molecule(Strandedness.SINGLE, MolType.RNA, Topology.LINEAR),
                [],
                438782,
            )


def test_create_isolate(empty_repo: EventSourcedRepo):
    """Test that creating an isolate returns the expected ``RepoIsolate`` object and
    creates the expected event file.
    """
    otu = init_otu(empty_repo)

    isolate = empty_repo.create_isolate(otu.id, None, "A", "isolate")

    assert isinstance(isolate.id, UUID)
    assert isolate.sequences == []
    assert isolate.name.value == "A"
    assert isolate.name.type == "isolate"

    with open(empty_repo.path.joinpath("src", "00000003.json")) as f:
        event = orjson.loads(f.read())

    del event["timestamp"]

    assert event == {
        "data": {
            "id": str(isolate.id),
            "legacy_id": None,
            "name": {"type": "isolate", "value": "A"},
        },
        "id": 3,
        "query": {
            "otu_id": str(otu.id),
            "isolate_id": str(isolate.id),
        },
        "type": "CreateIsolate",
    }

    assert empty_repo.last_id == 3


def test_create_sequence(empty_repo: EventSourcedRepo):
    """Test that creating a sequence returns the expected ``RepoSequence`` object and
    creates the expected event file.
    """
    otu = init_otu(empty_repo)

    isolate = empty_repo.create_isolate(otu.id, None, "A", "isolate")

    sequence = empty_repo.create_sequence(
        otu.id,
        isolate.id,
        "TMVABC",
        "TMV",
        None,
        "RNA",
        "ACGT",
    )

    assert sequence == EventSourcedRepoSequence(
        id=sequence.id,
        accession="TMVABC",
        definition="TMV",
        legacy_id=None,
        segment="RNA",
        sequence="ACGT",
    )

    with open(empty_repo.path.joinpath("src", "00000004.json")) as f:
        event = orjson.loads(f.read())

    del event["timestamp"]

    assert event == {
        "data": {
            "id": str(sequence.id),
            "accession": "TMVABC",
            "definition": "TMV",
            "legacy_id": None,
            "segment": "RNA",
            "sequence": "ACGT",
        },
        "id": 4,
        "query": {
            "otu_id": str(otu.id),
            "isolate_id": str(isolate.id),
            "sequence_id": str(sequence.id),
        },
        "type": "CreateSequence",
    }

    assert empty_repo.last_id == 4


class TestRetrieveOTU:
    def test_get_otu(self, empty_repo: EventSourcedRepo):
        """Test that getting an OTU returns the expected ``RepoOTU`` object including two
        isolates with one sequence each.
        """
        otu = empty_repo.create_otu(
            "TMV",
            None,
            "Tobacco mosaic virus",
            Molecule(Strandedness.SINGLE, MolType.RNA, Topology.LINEAR),
            [],
            12242,
        )

        isolate_a = empty_repo.create_isolate(otu.id, None, "A", "isolate")
        empty_repo.create_sequence(
            otu.id,
            isolate_a.id,
            "TMVABC",
            "TMV",
            None,
            "RNA",
            "ACGT",
        )

        isolate_b = empty_repo.create_isolate(otu.id, None, "B", "isolate")
        empty_repo.create_sequence(
            otu.id,
            isolate_b.id,
            "TMVABCB",
            "TMV",
            None,
            "RNA",
            "ACGTGGAGAGACC",
        )

        otu = empty_repo.get_otu(otu.id, ignore_cache=True)

        assert otu == EventSourcedRepoOTU(
            id=otu.id,
            acronym="TMV",
            excluded_accessions=set(),
            legacy_id=None,
            isolates=[
                EventSourcedRepoIsolate(
                    id=isolate_a.id,
                    legacy_id=None,
                    name=IsolateName(type="isolate", value="A"),
                    _sequences_by_accession={
                        "TMVABC": EventSourcedRepoSequence(
                            id=otu.isolates[0].sequences[0].id,
                            accession="TMVABC",
                            definition="TMV",
                            legacy_id=None,
                            segment="RNA",
                            sequence="ACGT",
                        ),
                    },
                ),
                EventSourcedRepoIsolate(
                    id=isolate_b.id,
                    legacy_id=None,
                    name=IsolateName(type="isolate", value="B"),
                    _sequences_by_accession={
                        "TMVABCB": EventSourcedRepoSequence(
                            id=otu.isolates[1].sequences[0].id,
                            accession="TMVABCB",
                            definition="TMV",
                            legacy_id=None,
                            segment="RNA",
                            sequence="ACGTGGAGAGACC",
                        ),
                    },
                ),
            ],
            name="Tobacco mosaic virus",
            molecule=Molecule(Strandedness.SINGLE, MolType.RNA, Topology.LINEAR),
            schema=[],
            taxid=12242,
        )

        assert empty_repo.last_id == 6

    def test_get_accessions(self, initialized_repo: EventSourcedRepo):
        otu = list(initialized_repo.iter_otus())[0]

        assert otu.accession_set == {"TMVABC"}

        isolate_b = initialized_repo.create_isolate(otu.id, None, "B", "isolate")
        initialized_repo.create_sequence(
            otu.id,
            isolate_b.id,
            "TMVABCB",
            "TMV",
            None,
            "RNA",
            "ACGTGGAGAGACC",
        )

        otu = list(initialized_repo.iter_otus())[0]

        assert otu.accession_set == {"TMVABC", "TMVABCB"}

    def test_get_blocked_accessions(self, initialized_repo: EventSourcedRepo):
        otu_id = initialized_repo.index_otus()[12242]
        isolate_b = initialized_repo.create_isolate(otu_id, None, "B", "isolate")
        initialized_repo.create_sequence(
            otu_id,
            isolate_b.id,
            "TMVABCB",
            "TMV",
            None,
            "RNA",
            "ACGTGGAGAGACC",
        )

        initialized_repo.exclude_accession(otu_id, "GROK")
        initialized_repo.exclude_accession(otu_id, "TOK")

        otu = initialized_repo.get_otu(otu_id)

        assert otu.blocked_accession_set == {"TMVABC", "TMVABCB", "GROK", "TOK"}

    def test_get_isolate(self, initialized_repo: EventSourcedRepo):
        otu = list(initialized_repo.iter_otus())[0]

        isolate_ids = {isolate.id for isolate in otu.isolates}

        for isolate_id in isolate_ids:
            assert otu.get_isolate(isolate_id) in otu.isolates

    def test_get_isolate_id_by_name(self, initialized_repo: EventSourcedRepo):
        otu = list(initialized_repo.iter_otus())[0]

        isolate_ids = {isolate.id for isolate in otu.isolates}

        assert (
            otu.get_isolate_id_by_name(IsolateName(type="isolate", value="A"))
            in isolate_ids
        )


def test_exclude_accession(empty_repo: EventSourcedRepo):
    """Test that excluding an accession from an OTU writes the expected event file and
    returns the expected OTU objects.
    """
    otu = empty_repo.create_otu(
        "TMV",
        None,
        "Tobacco mosaic virus",
        Molecule(Strandedness.SINGLE, MolType.RNA, Topology.LINEAR),
        [],
        12242,
    )

    empty_repo.exclude_accession(otu.id, "TMVABC")

    with open(empty_repo.path.joinpath("src", "00000003.json")) as f:
        event = orjson.loads(f.read())

        del event["timestamp"]

        assert event == {
            "data": {
                "accession": "TMVABC",
            },
            "id": 3,
            "query": {
                "otu_id": str(otu.id),
            },
            "type": "ExcludeAccession",
        }

    assert empty_repo.get_otu(otu.id, ignore_cache=True).excluded_accessions == {
        "TMVABC"
    }

    empty_repo.exclude_accession(otu.id, "ABTV")

    assert empty_repo.get_otu(otu.id, ignore_cache=True).excluded_accessions == {
        "TMVABC",
        "ABTV",
    }


class TestEventIndexCache:
    def test_equivalence(self, initialized_repo: EventSourcedRepo):
        assert initialized_repo.last_id == 4

        otu = list(initialized_repo.iter_otus())[0]

        otu_from_cache = initialized_repo.get_otu(otu.id, ignore_cache=False)
        otu_from_scratch = initialized_repo.get_otu(otu.id, ignore_cache=True)
        assert otu_from_cache == otu_from_scratch

        isolate_b = initialized_repo.create_isolate(otu.id, None, "B", "isolate")
        initialized_repo.create_sequence(
            otu.id,
            isolate_b.id,
            "TMVABCB",
            "TMV",
            None,
            "RNA",
            "ACGTGGAGAGACC",
        )

        assert initialized_repo.last_id == 6

        otu_from_cache = initialized_repo.get_otu(otu.id, ignore_cache=False)
        otu_from_scratch = initialized_repo.get_otu(otu.id, ignore_cache=True)
        assert otu_from_cache == otu_from_scratch

    def test_retrieve_nonexistent_otu(self, initialized_repo: EventSourcedRepo):
        assert (
            initialized_repo.get_otu(
                UUID("48b453fb-b60a-4f53-85f1-2941cd3ac5af"), ignore_cache=False
            )
            is None
        )

    def test_get_otu_without_cached_events(self, initialized_repo: EventSourcedRepo):
        otu = list(initialized_repo.iter_otus())[0]

        initialized_repo._event_index_cache.clear_cached_otu_events(otu.id)

        assert initialized_repo.get_otu(otu.id, ignore_cache=False) is not None
