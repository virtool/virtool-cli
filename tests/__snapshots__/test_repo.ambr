# serializer version: 1
# name: TestGetOTUByID.test_ok[otu]
  dict({
    'abbreviation': '',
    'exclusions': list([
    ]),
    'id': '3962f6ec',
    'isolates': list([
      dict({
        'default': True,
        'id': 'dbf54ede',
        'sequences': list([
        ]),
        'source_name': 'Mexico',
        'source_type': 'isolate',
      }),
    ]),
    'name': 'Jacquemontia mosaic Yucatan virus',
    'schema': list([
      dict({
        'molecule': 'ssDNA',
        'name': 'DNA A',
        'required': True,
      }),
      dict({
        'molecule': 'ssDNA',
        'name': 'DNA B',
        'required': True,
      }),
    ]),
    'taxid': 1198450,
  })
# ---
# name: TestGetOTUByTaxid.test_ok[otu]
  dict({
    'abbreviation': 'DLVd',
    'exclusions': list([
    ]),
    'id': 'ab6bd88a',
    'isolates': list([
      dict({
        'default': True,
        'id': 'fwth8bdi',
        'sequences': list([
        ]),
        'source_name': '4706174',
        'source_type': 'isolate',
      }),
    ]),
    'name': 'Dahlia latent viroid',
    'schema': list([
      dict({
        'molecule': 'ssRNA+',
        'name': 'Viroid',
        'required': True,
      }),
    ]),
    'taxid': 1278205,
  })
# ---
# name: test_create_otu[otu]
  dict({
    'abbreviation': '',
    'exclusions': list([
    ]),
    'isolates': list([
    ]),
    'name': '',
    'schema': list([
    ]),
    'taxid': -1,
  })
# ---
# name: test_init[otu_id_to_path]
  dict({
    '0bfdb8bc': 'wasabi_mottle_virus--0bfdb8bc',
    '3962f6ec': 'jacquemontia_mosaic_yucatan_virus--3962f6ec',
    '407d8a21': 'kalanchoe_latent_virus--407d8a21',
    '4c9ddfb7': 'lagenaria_siceraria_endornavirus_california--4c9ddfb7',
    '5aa4a083': 'east_african_cassava_mosaic_cameroon_virus--5aa4a083',
    '5d877144': 'yacon_necrotic_mottle_virus--5d877144',
    '6444acf3': 'faba_bean_necrotic_stunt_alphasatellite_1--6444acf3',
    '6bc81904': 'saccharum_streak_virus--6bc81904',
    '71d15b86': 'ugandan_cassava_brown_streak_virus--71d15b86',
    '9da7c10d': 'radish_leaf_curl_betasatellite--9da7c10d',
    'a89b6529': 'habenaria_mosaic_virus--a89b6529',
    'ab6bd88a': 'dahlia_latent_viroid--ab6bd88a',
    'ae0f2a35': 'nanovirus_like_particle--ae0f2a35',
    'c93ec9a9': 'abaca_bunchy_top_virus--c93ec9a9',
    'd226290f': 'cabbage_leaf_curl_jamaica_virus--d226290f',
    'd86d0412': 'tall_oatgrass_mosaic_virus--d86d0412',
    'dd21fd8f': 'pagoda_yellow_mosaic_associated_virus--dd21fd8f',
    'e1dfb3ef': 'oat_blue_dwarf_virus--e1dfb3ef',
    'e5b6dd57': 'zantedeschia_mild_mosaic_virus--e5b6dd57',
    'e7wkndjc': 'impatiens_flower_break_potyvirus--e7wkndjc',
    'f8a56910': 'gaillardia_latent_virus--f8a56910',
    'g7qy3n72': 'maclura_mosaic_virus--g7qy3n72',
    'v7r8gxav': 'vallota_mosaic_virus--v7r8gxav',
    'xcl20vqt': 'babaco_mosaic_virus--xcl20vqt',
  })
# ---
# name: test_init[sequence_id_to_otu_id]
  dict({
    '0owj7oas': 'c93ec9a9',
    '0s9kczxs': 'f8a56910',
    '27ficy1s': 'dd21fd8f',
    '5ms4tpsb': 'd86d0412',
    '5w3oz9pl': 'c93ec9a9',
    '6yy1d2ev': '5aa4a083',
    'atfrq1dt': 'a89b6529',
    'buvquaqv': 'e5b6dd57',
    'clt0zobl': '0bfdb8bc',
    'cpm1g5h4': 'c93ec9a9',
    'd9bbetey': '0bfdb8bc',
    'e5lz6r5m': '6bc81904',
    'emrw6o2d': 'd226290f',
    'fgdkm0kp': 'c93ec9a9',
    'fl6r8ylz': 'c93ec9a9',
    'ggx1bvx6': 'e1dfb3ef',
    'gxansv52': '0bfdb8bc',
    'heprpw4w': 'c93ec9a9',
    'ib2dn2kt': 'xcl20vqt',
    'iz8ant3f': 'c93ec9a9',
    'j0g7sfp4': 'c93ec9a9',
    'jj1jtoj0': '9da7c10d',
    'jowusser': '5d877144',
    'm7y38816': '0bfdb8bc',
    'ndaxyl7f': 'd226290f',
    'ngn9uhrp': '5aa4a083',
    'ondpjl1h': '0bfdb8bc',
    'pqm6etvy': 'c93ec9a9',
    'pvx374vw': 'v7r8gxav',
    'pyb0vqd8': 'd226290f',
    'pyf9x4wh': 'd226290f',
    'qrhnfccf': '6444acf3',
    's86vn1ie': '3962f6ec',
    't2ppu39d': 'c93ec9a9',
    'ti04jjqb': 'e7wkndjc',
    'tw0nwtro': '4c9ddfb7',
    'usvsrzjk': '407d8a21',
    'v97n3owu': 'g7qy3n72',
    'vza1fxor': 'c93ec9a9',
    'wuf8wjhh': '3962f6ec',
    'wv62fdpe': '71d15b86',
    'wzpj56v7': 'c93ec9a9',
    'x1ktdmdu': 'ab6bd88a',
    'yf43abg4': 'ae0f2a35',
  })
# ---
# name: test_init[sequence_id_to_path]
  dict({
    '0owj7oas': '0owj7oas.json',
    '0s9kczxs': '0s9kczxs.json',
    '27ficy1s': '27ficy1s.json',
    '5ms4tpsb': '5ms4tpsb.json',
    '5w3oz9pl': '5w3oz9pl.json',
    '6yy1d2ev': '6yy1d2ev.json',
    'atfrq1dt': 'atfrq1dt.json',
    'buvquaqv': 'buvquaqv.json',
    'clt0zobl': 'clt0zobl.json',
    'cpm1g5h4': 'cpm1g5h4.json',
    'd9bbetey': 'd9bbetey.json',
    'e5lz6r5m': 'e5lz6r5m.json',
    'emrw6o2d': 'emrw6o2d.json',
    'fgdkm0kp': 'fgdkm0kp.json',
    'fl6r8ylz': 'fl6r8ylz.json',
    'ggx1bvx6': 'ggx1bvx6.json',
    'gxansv52': 'gxansv52.json',
    'heprpw4w': 'heprpw4w.json',
    'ib2dn2kt': 'ib2dn2kt.json',
    'iz8ant3f': 'iz8ant3f.json',
    'j0g7sfp4': 'j0g7sfp4.json',
    'jj1jtoj0': 'jj1jtoj0.json',
    'jowusser': 'jowusser.json',
    'm7y38816': 'm7y38816.json',
    'ndaxyl7f': 'ndaxyl7f.json',
    'ngn9uhrp': 'ngn9uhrp.json',
    'ondpjl1h': 'ondpjl1h.json',
    'pqm6etvy': 'pqm6etvy.json',
    'pvx374vw': 'pvx374vw.json',
    'pyb0vqd8': 'pyb0vqd8.json',
    'pyf9x4wh': 'pyf9x4wh.json',
    'qrhnfccf': 'qrhnfccf.json',
    's86vn1ie': 's86vn1ie.json',
    't2ppu39d': 't2ppu39d.json',
    'ti04jjqb': 'ti04jjqb.json',
    'tw0nwtro': 'tw0nwtro.json',
    'usvsrzjk': 'usvsrzjk.json',
    'v97n3owu': 'v97n3owu.json',
    'vza1fxor': 'vza1fxor.json',
    'wuf8wjhh': 'wuf8wjhh.json',
    'wv62fdpe': 'wv62fdpe.json',
    'wzpj56v7': 'wzpj56v7.json',
    'x1ktdmdu': 'x1ktdmdu.json',
    'yf43abg4': 'yf43abg4.json',
  })
# ---
# name: test_init[taxid_to_otu_id]
  dict({
    48201: 'g7qy3n72',
    56879: 'e1dfb3ef',
    96892: 'ae0f2a35',
    132477: '407d8a21',
    223262: '5aa4a083',
    270478: 'e5b6dd57',
    345184: 'd226290f',
    430059: 'v7r8gxav',
    438782: 'c93ec9a9',
    662596: '9da7c10d',
    683179: '6bc81904',
    946046: '71d15b86',
    1169032: '0bfdb8bc',
    1198450: '3962f6ec',
    1278205: 'ab6bd88a',
    1324128: 'a89b6529',
    1414644: 'd86d0412',
    1441799: '6444acf3',
    1468172: 'f8a56910',
    1468454: '4c9ddfb7',
    1505530: 'dd21fd8f',
    1561150: '5d877144',
    2060511: 'xcl20vqt',
    2170135: 'e7wkndjc',
  })
# ---
